spec-abstract-noclosure/ffg.tla

```diff
@@ -0,0 +1,337 @@
+    /\ ffg_votes' = ffg_votes \union { ffgVote }
+    /\ votes' = votes \union { Vote(v, ffgVote): v \in validators }
+    /\ LET allCheckpoints == { Checkpoint(block, i): block \in all_blocks, i \in CheckpointSlots } IN
+       \E allJustifiedCheckpoints \in SUBSET allCheckpoints:
+        /\ justified_checkpoints' = allJustifiedCheckpoints
+        /\ \A c \in allJustifiedCheckpoints: IsJustified(c, votes', allJustifiedCheckpoints)
+        /\ \A c \in (allCheckpoints \ allJustifiedCheckpoints): ~IsJustified(c, votes', allJustifiedCheckpoints)
+    \*/\ justified_checkpoints' = JustifiedCheckpoints(votes')
+    /\ UNCHANGED <<all_blocks, chain1, chain1_tip, chain2, chain2_tip, chain2_fork_block_number>>
+
+\* Check this property to see an example of two conflicting blocks.
+TwoConflictingBlocks ==
+    \A b1, b2 \in all_blocks: ~AreConflictingBlocks(b1, b2)
+
+(*
+ Check this property to see an example of two finalized conflicting blocks.
+
+ Scenario 1:
+
+ slot 0  slot 1  slot 2  slot 3  slot 4  slot 5
+
+ genesis  b1      c1     c1 [f]  
+         -b1                       -c1   -c1 [f]
+ *)
+TwoFinalizedConflictingBlocks ==
+    LET disagreement == \E c1, c2 \in justified_checkpoints: 
+        /\ IsFinalized(c1, votes, justified_checkpoints)
+        /\ IsFinalized(c2, votes, justified_checkpoints)
+        /\ AreConflictingBlocks(c1[1], c2[1])
+    IN ~disagreement
+
+(*
+ Check this property to see an example of two finalized conflicting blocks,
+ which are not caused by equivocation.
+
+ Scenario 2:
+
+ slot 0    slot 1    slot 2   slot 3    slot 4  slot 5   slot 6   slot 7
+                        
+ genesis    b1         c1
+                       b2                 c2     c2 [f]
+            -b1        -b2     -c1                         -c2     -c2 [f]
+ *)
+TwoFinalizedConflictingBlocksNoEquivocation ==
+    LET \* @type: ($vote, $vote) => Bool;
+        IsEquivocation(vote1, vote2) ==
+        /\ vote1 /= vote2
+        /\ vote1.ffg_vote.target[2] = vote2.ffg_vote.target[2]
+    IN
+    LET disagreement == \E c1, c2 \in justified_checkpoints: 
+        /\ IsFinalized(c1, votes, justified_checkpoints)
+        /\ IsFinalized(c2, votes, justified_checkpoints)
+        /\ AreConflictingBlocks(c1[1], c2[1])
+    IN \/ ~disagreement
+       \/ \E vote1, vote2 \in votes:
+            /\ vote1.validator = vote2.validator
+            /\ IsEquivocation(vote1, vote2)
+
+AccountableSafety ==
+    \/ SlashableNodesOver \*Cardinality(SlashableNodes) * 3 >= N
+    \/ \A c1, c2 \in justified_checkpoints:
+        \/ ~IsFinalized(c1, votes, justified_checkpoints)
+        \/ ~IsFinalized(c2, votes, justified_checkpoints)
+        \/ ~AreConflictingBlocks(c1[1], c2[1])
+
+Init == 
+    /\ all_blocks = { GenesisBlock }
+    /\ chain1_tip = GenesisBlock
+    /\ chain2_tip = GenesisBlock
+    /\ chain1 = { GenesisBlock }
+    /\ chain2 = { GenesisBlock }
+    /\ chain2_fork_block_number = 0
+    /\ ffg_votes = {}
+    /\ votes = {}
```

Comment:
Instead of a set, we can encode `votes` as a function `votes = [ v \in VALIDATORS |-> {} ] ` where `votes[v]` is a set that contains only votes from validator `v`.

By using this encoding, we can rewrite `IsJustified` as the following condition. The only difference is that Apalache needs to look for `justifyingVote` in a set of messages from `v` that is stored in `viewVotes[v]`.
This change may be helpful for Apalache.

```
IsJustified(checkpoint, viewVotes, fixpoint) == 
    \/ checkpoint = GenesisCheckpoint
    \/ LET validatorsWhoCastJustifyingVote == { 
        v \in VALIDATORS: \E justifyingVote \in viewVotes[v]:  \* Look for justifyingVote in messages only from v
            /\ justifyingVote.validator = v
            /\ LET ffgVote == justifyingVote.ffg_vote IN
                \* L6:
                /\ ffgVote.source \in fixpoint
                \* L7:
                /\ IsLeftAncestorOfRight(checkpoint[1], ffgVote.target[1])
                /\ IsLeftAncestorOfRight(ffgVote.source[1], checkpoint[1])
                \* L8:
                /\ ffgVote.target[2] = checkpoint[2] }    
        IN validatorsWhoCastJustifyingVote \in LargeQuorums
```

Reply:
We can also consider the following way:

```
IsJustified(checkpoint, viewVotes, fixpoint) == 
    \/ checkpoint = GenesisCheckpoint
    \/ \E quorum \in LargeQuorums : \A v \in quorum : \E justifyingVote \in viewVotes[v] : 
            /\ justifyingVote.validator = v
            /\ LET ffgVote == justifyingVote.ffg_vote IN
                \* L6:
                /\ ffgVote.source \in fixpoint
                \* L7:
                /\ IsLeftAncestorOfRight(checkpoint[1], ffgVote.target[1])
                /\ IsLeftAncestorOfRight(ffgVote.source[1], checkpoint[1])
                \* L8:
                /\ ffgVote.target[2] = checkpoint[2] 
```

While this encoding requires more quantifiers, it may be also good for Apalache as only one universal quantifier is used and others are existential quantifiers.

Reply:
Those encoding can be also applied to `IsFinalized`.

Reply:
I tried it with a function in https://github.com/freespek/ssf-mc/tree/igor/rethink-blocks-votes

It's hard to tell whether this brings any improvement, as we still have non-terminating queries.